File: Shopify/pair_programming/robot_extensible_commands.cpp
/*
 * Shopify Pair Programming - Robot with Extensible Command System
 * Confirmed: Dec 2025 (PracHub, MLE Technical Screen)
 *
 * This is a VARIANT of the basic robot grid problem.
 * The KEY DIFFERENCE: the problem spec explicitly tells you
 * "valid commands will keep expanding" (examples given: backward, jump, undo).
 *
 * This means the interviewer is testing whether you will:
 *   a) Hard-code if/else chains (wrong — doesn't scale)
 *   b) Design a command registry / command-object pattern (correct)
 *
 * Staff-level expectation: you should proactively propose the extensible
 * design BEFORE being asked, not wait for the interviewer to say "now add undo."
 *
 * Problem:
 *   Robot starts at (x0, y0, dir0).
 *   Execute a string of commands. Return final state.
 *   Commands: F (forward), L (turn left), R (turn right)
 *   More commands WILL be added.
 *
 * Layer 2: Invalid command policy (ignore / raise / collect failures)
 * Layer 3: Testing discussion — unit, property-based, edge cases
 * Layer 4: Add UNDO command (undo last non-UNDO move)
 */

#include <iostream>
#include <string>
#include <unordered_map>
#include <functional>
#include <stdexcept>
#include <vector>
#include <stack>
#include <cassert>

// ============================================================
// Core types
// ============================================================

enum class Direction { NORTH = 0, EAST = 1, SOUTH = 2, WEST = 3 };

std::string dirStr(Direction d) {
    switch (d) {
        case Direction::NORTH: return "N";
        case Direction::EAST:  return "E";
        case Direction::SOUTH: return "S";
        case Direction::WEST:  return "W";
    }
    return "?";
}

struct State {
    int x, y;
    Direction dir;

    bool operator==(const State& o) const {
        return x == o.x && y == o.y && dir == o.dir;
    }
};

// ============================================================
// Command registry: new commands are registered, not hard-coded
// ============================================================

using CommandFn = std::function<State(State)>;

class CommandRegistry {
public:
    void registerCommand(char symbol, CommandFn fn) {
        commands[symbol] = std::move(fn);
    }

    bool has(char symbol) const { return commands.count(symbol) > 0; }

    State execute(char symbol, State state) const {
        return commands.at(symbol)(state);
    }

private:
    std::unordered_map<char, CommandFn> commands;
};

// ============================================================
// Default commands
// ============================================================

CommandRegistry makeDefaultRegistry() {
    CommandRegistry reg;

    reg.registerCommand('F', [](State s) {
        switch (s.dir) {
            case Direction::NORTH: s.y++; break;
            case Direction::EAST:  s.x++; break;
            case Direction::SOUTH: s.y--; break;
            case Direction::WEST:  s.x--; break;
        }
        return s;
    });

    reg.registerCommand('L', [](State s) {
        s.dir = static_cast<Direction>((static_cast<int>(s.dir) + 3) % 4);
        return s;
    });

    reg.registerCommand('R', [](State s) {
        s.dir = static_cast<Direction>((static_cast<int>(s.dir) + 1) % 4);
        return s;
    });

    // Layer 4: Backward command (easy to add without changing anything else)
    reg.registerCommand('B', [](State s) {
        switch (s.dir) {
            case Direction::NORTH: s.y--; break;
            case Direction::EAST:  s.x--; break;
            case Direction::SOUTH: s.y++; break;
            case Direction::WEST:  s.x++; break;
        }
        return s;
    });

    return reg;
}

// ============================================================
// Robot executor with configurable invalid-command policy
// ============================================================

enum class InvalidPolicy { IGNORE, RAISE, COLLECT };

struct ExecutionResult {
    State finalState;
    std::vector<char> invalidCommands; // only populated with COLLECT policy
};

ExecutionResult execute(
    const std::string& commands,
    State initial,
    const CommandRegistry& registry,
    InvalidPolicy policy = InvalidPolicy::IGNORE)
{
    State state = initial;
    std::vector<char> invalid;

    for (char c : commands) {
        if (!registry.has(c)) {
            if (policy == InvalidPolicy::RAISE) {
                throw std::invalid_argument(std::string("Unknown command: ") + c);
            } else if (policy == InvalidPolicy::COLLECT) {
                invalid.push_back(c);
            }
            // IGNORE: do nothing
            continue;
        }
        state = registry.execute(c, state);
    }

    return {state, invalid};
}

// ============================================================
// Layer 4: UNDO support via history stack
// ============================================================

class RobotWithUndo {
public:
    explicit RobotWithUndo(State initial, CommandRegistry registry)
        : state(initial), registry(std::move(registry)) {}

    void execute(char c) {
        if (c == 'U') {
            undo();
            return;
        }
        if (!registry.has(c)) return; // IGNORE policy
        history.push(state);
        state = registry.execute(c, state);
    }

    void execute(const std::string& commands) {
        for (char c : commands) execute(c);
    }

    State getState() const { return state; }

private:
    State state;
    CommandRegistry registry;
    std::stack<State> history;

    void undo() {
        if (history.empty()) return; // nothing to undo
        state = history.top();
        history.pop();
    }
};

// ============================================================
// Tests
// ============================================================

void runTests() {
    auto registry = makeDefaultRegistry();

    // Basic movement
    {
        auto [s, _] = execute("FFRFF", {0, 0, Direction::NORTH}, registry);
        assert(s.x == 2 && s.y == 2);
        std::cout << "FFRFF: (" << s.x << "," << s.y << "," << dirStr(s.dir) << ") PASS" << std::endl;
    }

    // Invalid command — ignore policy
    {
        auto [s, inv] = execute("FXF", {0, 0, Direction::NORTH}, registry, InvalidPolicy::IGNORE);
        assert(s.y == 2); // X ignored, 2 forward steps
        std::cout << "Ignore invalid: PASS" << std::endl;
    }

    // Invalid command — collect policy
    {
        auto [s, inv] = execute("FXF", {0, 0, Direction::NORTH}, registry, InvalidPolicy::COLLECT);
        assert(inv.size() == 1 && inv[0] == 'X');
        std::cout << "Collect invalid: PASS (" << inv.size() << " invalid)" << std::endl;
    }

    // Invalid command — raise policy
    {
        try {
            execute("FXF", {0, 0, Direction::NORTH}, registry, InvalidPolicy::RAISE);
            std::cout << "FAIL: should have thrown" << std::endl;
        } catch (const std::invalid_argument& e) {
            std::cout << "Raise on invalid: PASS" << std::endl;
        }
    }

    // Backward command (dynamically registered, no core changes)
    {
        auto [s, _] = execute("FB", {0, 0, Direction::NORTH}, registry);
        assert(s.y == 0); // F up 1, B down 1 = net 0
        std::cout << "Forward + Backward: PASS (" << s.x << "," << s.y << ")" << std::endl;
    }

    // Undo
    {
        RobotWithUndo robot({0, 0, Direction::NORTH}, makeDefaultRegistry());
        robot.execute("FF");      // (0,2,N)
        robot.execute("U");       // undo -> (0,1,N)
        robot.execute("R");       // (0,1,E)
        robot.execute("U");       // undo -> (0,1,N)
        auto s = robot.getState();
        assert(s.x == 0 && s.y == 1 && s.dir == Direction::NORTH);
        std::cout << "Undo test: (" << s.x << "," << s.y << "," << dirStr(s.dir) << ") PASS" << std::endl;
    }
}

int main() {
    runTests();
    return 0;
}

File: Shopify/java/RobotExtensibleCommands.java
import java.util.*;
import java.util.function.UnaryOperator;

/**
 * Shopify Pair Programming - Robot with an Extensible Command System
 * Confirmed: Dec 2025 (PracHub, MLE Technical Screen).
 *
 * Java 11.  Run:  javac RobotExtensibleCommands.java && java RobotExtensibleCommands
 *
 * KEY: the spec explicitly says "valid commands will keep expanding"
 * (backward, jump, undo...). The whole point is the Open/Closed Principle:
 * add a command by REGISTERING it, never by editing the executor's switch.
 *
 * SOLID, not over-engineered: a Command is just State -> State. No class
 * explosion, no visitor, no DI container - just a map of small functions.
 */
public class RobotExtensibleCommands {

    enum Direction {
        NORTH(0, 1), EAST(1, 0), SOUTH(0, -1), WEST(-1, 0);
        final int dx, dy;
        Direction(int dx, int dy) { this.dx = dx; this.dy = dy; }
        Direction left()  { return values()[(ordinal() + 3) % 4]; }
        Direction right() { return values()[(ordinal() + 1) % 4]; }
    }

    static final class State {
        private final int x, y;
        private final Direction dir;
        State(int x, int y, Direction dir) { this.x = x; this.y = y; this.dir = dir; }
        int x() { return x; }
        int y() { return y; }
        Direction dir() { return dir; }
        @Override public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof State)) return false;
            State s = (State) o;
            return x == s.x && y == s.y && dir == s.dir;
        }
        @Override public int hashCode() { return Objects.hash(x, y, dir); }
        @Override public String toString() { return "(" + x + "," + y + "," + dir + ")"; }
    }

    /** A command is a pure State transformation. Named for readability. */
    @FunctionalInterface
    interface Command extends UnaryOperator<State> {}

    /** Open for extension (register), closed for modification (no switch to edit). */
    static class CommandRegistry {
        private final Map<Character, Command> commands = new HashMap<>();
        CommandRegistry register(char symbol, Command cmd) { commands.put(symbol, cmd); return this; }
        boolean has(char symbol) { return commands.containsKey(symbol); }
        State apply(char symbol, State s) { return commands.get(symbol).apply(s); }
    }

    enum InvalidPolicy { IGNORE, RAISE, COLLECT }

    static final class ExecutionResult {
        private final State finalState;
        private final List<Character> invalid;
        ExecutionResult(State finalState, List<Character> invalid) {
            this.finalState = finalState; this.invalid = invalid;
        }
        State finalState() { return finalState; }
        List<Character> invalid() { return invalid; }
    }

    static CommandRegistry defaultRegistry() {
        return new CommandRegistry()
            .register('F', s -> new State(s.x() + s.dir().dx, s.y() + s.dir().dy, s.dir()))
            .register('B', s -> new State(s.x() - s.dir().dx, s.y() - s.dir().dy, s.dir()))
            .register('L', s -> new State(s.x(), s.y(), s.dir().left()))
            .register('R', s -> new State(s.x(), s.y(), s.dir().right()));
    }

    static ExecutionResult execute(String commands, State initial,
                                   CommandRegistry registry, InvalidPolicy policy) {
        State state = initial;
        List<Character> invalid = new ArrayList<>();
        for (char c : commands.toCharArray()) {
            if (!registry.has(c)) {
                switch (policy) {
                    case RAISE:   throw new IllegalArgumentException("Unknown command: " + c);
                    case COLLECT: invalid.add(c); break;
                    case IGNORE:  break;
                }
                continue;
            }
            state = registry.apply(c, state);
        }
        return new ExecutionResult(state, invalid);
    }

    /**
     * Layer 4: UNDO. Note UNDO is NOT a pure Command - it needs history, which
     * is executor state. Keeping it out of the registry is a deliberate boundary
     * (Single Responsibility): commands transform position; the executor owns history.
     */
    static class UndoableRobot {
        private final CommandRegistry registry;
        private final Deque<State> history = new ArrayDeque<>();
        private State state;

        UndoableRobot(State initial, CommandRegistry registry) {
            this.state = initial;
            this.registry = registry;
        }
        void execute(char c) {
            if (c == 'U') { undo(); return; }
            if (!registry.has(c)) return;          // IGNORE unknowns
            history.push(state);
            state = registry.apply(c, state);
        }
        void execute(String s) { for (char c : s.toCharArray()) execute(c); }
        private void undo() { if (!history.isEmpty()) state = history.pop(); }
        State state() { return state; }
    }

    public static void main(String[] args) {
        CommandRegistry reg = defaultRegistry();

        ExecutionResult r1 = execute("FFRFF", new State(0, 0, Direction.NORTH), reg, InvalidPolicy.IGNORE);
        check("Basic", r1.finalState().equals(new State(2, 2, Direction.EAST)), r1.finalState().toString());

        ExecutionResult r2 = execute("FXF", new State(0, 0, Direction.NORTH), reg, InvalidPolicy.COLLECT);
        check("Collect", r2.invalid().equals(Collections.singletonList('X')) && r2.finalState().y() == 2,
              r2.invalid().toString());

        try {
            execute("FXF", new State(0, 0, Direction.NORTH), reg, InvalidPolicy.RAISE);
            check("Raise", false, "no throw");
        } catch (IllegalArgumentException e) {
            check("Raise", true, e.getMessage());
        }

        // Extend at RUNTIME with a brand-new 'J' (jump 2) command - executor unchanged.
        reg.register('J', s -> new State(s.x() + 2 * s.dir().dx, s.y() + 2 * s.dir().dy, s.dir()));
        ExecutionResult r3 = execute("J", new State(0, 0, Direction.NORTH), reg, InvalidPolicy.IGNORE);
        check("Extend(J)", r3.finalState().equals(new State(0, 2, Direction.NORTH)), r3.finalState().toString());

        UndoableRobot robot = new UndoableRobot(new State(0, 0, Direction.NORTH), defaultRegistry());
        robot.execute("FF");  // (0,2,N)
        robot.execute("U");   // (0,1,N)
        robot.execute("R");   // (0,1,E)
        robot.execute("U");   // (0,1,N)
        check("Undo", robot.state().equals(new State(0, 1, Direction.NORTH)), robot.state().toString());
    }

    static void check(String name, boolean pass, String actual) {
        System.out.printf("%-10s %s  (%s)%n", name, pass ? "PASS" : "FAIL", actual);
    }
}
